main.go

```diff
@@ -0,0 +1,200 @@
+)
+
+var version = "undefined"
+
+// System is a type that will be exported as an RPC service.
+type System int
+
+type Args struct{}
+
+// SystemResponse holds the result of the Multiply method.
+type SystemResponse struct {
+	Up int
+}
+
+// nolint:unparam // for the first version this implementation is enough
+func (a *System) Up(_ *Args, reply *SystemResponse) error {
+	reply.Up = 1
+	return nil
+}
+
+type httpReadWriteCloser struct {
+	in  io.Reader
+	out io.Writer
+}
+
+func (h *httpReadWriteCloser) Read(p []byte) (n int, err error)  { return h.in.Read(p) }
+func (h *httpReadWriteCloser) Write(p []byte) (n int, err error) { return h.out.Write(p) }
+func (h *httpReadWriteCloser) Close() error                      { return nil }
+
+func main() {
+	if err := run(); err != nil {
+		slog.Error("error occurred", "err", err)
+	}
+}
+
+func run() error {
+	c := config.Load()
+
+	setLogger(c.Debug)
+	c.LogFields()
+
+	ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt, os.Kill, syscall.SIGTERM)
+	defer stop()
+
+	group, ctx := errgroup.WithContext(ctx)
+
+	group.Go(func() error {
+		cli, err := ethclient.Dial(c.Rpc)
+		if err != nil {
+			return fmt.Errorf("error instantiating eth client: %w", err)
+		}
+		s := scan.NewScanner(cli, common.HexToAddress(c.ContractAddress))
+		return runScan(ctx, *c, cli, s)
+	})
+
+	// Create an HTTP handler for RPC
+	handler := http.NewServeMux()
+	handler.HandleFunc("/rpc", func(w http.ResponseWriter, r *http.Request) {
+		serverCodec := jsonrpc.NewServerCodec(&httpReadWriteCloser{r.Body, w})
+		w.Header().Set("Content-type", "application/json")
+		rpcErr := rpc.ServeRequest(serverCodec)
+		if rpcErr != nil {
+			slog.Warn("error while serving JSON request", "err", rpcErr)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	})
+
+	// Create an HTTP server with timeouts
+	server := &http.Server{
+		Addr:         fmt.Sprintf(":%d", c.Port),
+		Handler:      handler,
+		ReadTimeout:  5 * time.Second,
+		WriteTimeout: 10 * time.Second,
+		IdleTimeout:  120 * time.Second,
+	}
+
+	slog.Info("starting universal node RPC server", "port", c.Port)
+
+	group.Go(func() error {
+		<-ctx.Done()
+		slog.Info("shutting down the RPC server...")
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		if shutdownErr := server.Shutdown(shutdownCtx); shutdownErr != nil {
+			return fmt.Errorf("error shutting down the RPC server: %w", shutdownErr)
+		}
+		return nil
+	})
+	group.Go(func() error {
+		sys := new(System)
+		err := rpc.Register(sys)
+		if err != nil {
+			return fmt.Errorf("error registering RPC service: %w", err)
+		}
+		if srvErr := server.ListenAndServe(); srvErr != nil && srvErr != http.ErrServerClosed {
+			return srvErr
+		}
+		return nil
+	})
+	if err := group.Wait(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func runScan(ctx context.Context, c config.Config, cli scan.EthClient, s scan.Scanner) error {
```

Comment:
could you add a test for this?

func TestRunScan(t *testing.T) {
	mockCtrl := gomock.NewController(t)
	defer mockCtrl.Finish()

	mockEthClient := mock.NewMockEthClient(mockCtrl)
	mockScanner := mock.NewMockScanner(mockCtrl)

	mockEthClient.EXPECT().BlockNumber(gomock.Any()).Return(uint64(100), nil).AnyTimes()
	mockScanner.EXPECT().ScanEvents(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, nil).AnyTimes()

	c := config.Config{
		StartingBlock: 0,
		BlocksRange:   100,
	}

	// Create a context with a timeout
	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
	defer cancel() 
	err := runScan(ctx, c, mockEthClient, mockScanner)
	if err != nil {
		t.Errorf("runScan() returned an error: %v", err)
	}
}